train-controller/src/main/java/hu/bme/mit/train/controller/TrainControllerImpl.java

```diff
@@ -7,6 +7,22 @@
 	private int step = 0;
 	private int referenceSpeed = 0;
 	private int speedLimit = 0;
+	private Thread thread;
+	public TrainControllerImpl() {
+		thread = new Thread(){
+			public void run()
+			{
+				thread.run();
+				try{
+					followSpeed();
+					thread.sleep(1);
```

Comment:
It is not necessary to refresh this frequently.
You should change the sleep time for 500ms or more but maximum 1 second.
Also do not use magic numbers in the code! Make it a constant.

Reply:
Refresh rate set to 500 and added REFRESH_RATE variable

Reply:
Refresh rate set to 500 and added REFRESH_RATE variable